packages/babel-core/src/config/build-config-chain.js

```diff
@@ -131,31 +152,34 @@ class ConfigChainBuilder {
 
   mergeConfig({
     type,
-    options,
+    options: rawOpts,
     alias,
-    loc,
     dirname,
   }) {
-    if (!options) {
-      return false;
-    }
-
     // Bail out ASAP if this file is ignored so that we run as little logic as possible on ignored files.
-    if (this.filename && this.shouldIgnore(options.ignore, options.only, dirname)) {
+    if (this.filename && this.shouldIgnore(rawOpts.ignore || null, rawOpts.only || null, dirname)) {
       // TODO(logan): This is a really cross way to bail out. Avoid this in rewrite.
-      throw Object.assign(new Error("This file has been ignored."), { code: "BABEL_IGNORED_FILE" });
+      throw Object.assign((new Error("This file has been ignored."): any), { code: "BABEL_IGNORED_FILE" });
     }
 
-    options = Object.assign({}, options);
-
-    loc = loc || alias;
+    const options = Object.assign({}, rawOpts);
+    delete options.env;
+    delete options.extends;
 
     // env
-    const envKey = babel.getEnv();
-    if (options.env) {
-      const envOpts = options.env[envKey];
-      delete options.env;
+    const envKey = getEnv();
 
+    if (rawOpts.env != null && (typeof rawOpts.env !== "object" || Array.isArray(rawOpts.env))) {
+      throw new Error(".env block must be an object, null, or undefined");
```

Comment:
I think we should use a helper function for this kind of error messages.
We can also clarify a bit the message:

```
The `env` key in your Babel configuration must be ..., ... given
```

And maybe link to the website?